crypto/fuzzer/fuzzer.c

```diff
@@ -96,82 +119,103 @@ void fuzzer_reset_state(void) {
+#else
+  (void)pointer;
+  (void)length;
+#endif
+#endif
+}
+
+// simplify the pointer check after a var_pointer = malloc()
+#define RETURN_IF_NULL(var_pointer) \
+  if (var_pointer == NULL) {        \
+    return 0;                       \
+  }
+
+void zkp_initialize_context_or_crash(void) {
+  // The current context usage has persistent side effects
+  // TODO switch to frequent re-initialization where necessary
+  if (!zkp_context_is_initialized()) {
+    if (zkp_context_init() != 0) {
+      crash();
+    }
+  }
+}
+
 /* individual fuzzer harness functions */
 
 int fuzz_bn_format(void) {
   bignum256 target_bignum;
-  // we need some amount of data, bail if the input is too short
-  if (fuzzer_length < sizeof(target_bignum)) {
-    return 0;
+  // we need some amount of initial data
+  if (fuzzer_length < sizeof(target_bignum) + 1 + 1) {
+    return -1;
   }
 
 #define FUZZ_BN_FORMAT_OUTPUT_BUFFER_SIZE 512
   char buf[FUZZ_BN_FORMAT_OUTPUT_BUFFER_SIZE] = {0};
-  int ret;
+  int ret = 0;
 
   // mutate the struct contents
-  memcpy(&target_bignum, fuzzer_ptr, sizeof(target_bignum));
-  fuzzer_input(sizeof(target_bignum));
+  memcpy(&target_bignum, fuzzer_input(sizeof(target_bignum)),
+         sizeof(target_bignum));
 
   uint8_t prefixlen = 0;
-  if (fuzzer_length < 1) {
-    return 0;
-  }
-  memcpy(&prefixlen, fuzzer_input(1), 1);
-  char prefix[prefixlen];
-  memset(&prefix, 0, prefixlen);
-
-  if (prefixlen > 0 && prefixlen <= 128 && prefixlen <= fuzzer_length) {
-    memcpy(&prefix, fuzzer_input(prefixlen), prefixlen);
-    // force null termination
-    prefix[prefixlen - 1] = 0;
-  } else {
-    return 0;
-  }
-  // TODO idea: : allow prefix == NULL
-
   uint8_t suffixlen = 0;
-  if (fuzzer_length < 1) {
-    return 0;
-  }
-  memcpy(&suffixlen, fuzzer_input(1), 1);
-  char suffix[suffixlen];
-  memset(&suffix, 0, suffixlen);
-
-  if (suffixlen > 0 && suffixlen <= 128 && suffixlen <= fuzzer_length) {
-    memcpy(&suffix, fuzzer_input(suffixlen), suffixlen);
-    // force null termination
-    suffix[suffixlen - 1] = 0;
-  } else {
-    return 0;
-  }
-  // TODO idea: allow suffix == NULL
   uint32_t decimals = 0;
   int32_t exponent = 0;
   bool trailing = false;
+  // range 1 to 128
+  prefixlen = (fuzzer_input(1)[0] & 127) + 1;
+  suffixlen = (fuzzer_input(1)[0] & 127) + 1;
 
-  if (fuzzer_length >= 9) {
-    memcpy(&decimals, fuzzer_input(4), 4);
-    memcpy(&exponent, fuzzer_input(4), 4);
-
-    trailing = (fuzzer_input(1)[0] & 1);
-  } else {
-    return 0;
+  // check for the second half of the data
+  if (fuzzer_length < prefixlen + suffixlen + 4 + 4 + 1 - 2) {
+    return -1;
   }
+  memcpy(&decimals, fuzzer_input(4), 4);
+  memcpy(&exponent, fuzzer_input(4), 4);
+  trailing = (fuzzer_input(1)[0] & 1);
+
+  // IDEA allow prefix == NULL
+  char *prefix = malloc(prefixlen);
+  // IDEA allow suffix == NULL
+  char *suffix = malloc(suffixlen);
+  RETURN_IF_NULL(prefix);
+  RETURN_IF_NULL(suffix);
```

Comment:
Technically, if the second `malloc()` fails, then we should free `prefix`.

Reply:
Good catch. I've switched to a different ordering of `malloc` and `RETURN_IF_NULL` for code positions with more than one allocation to account for this.

Reply:
Resolved in https://github.com/trezor/trezor-firmware/pull/2449/commits/66db14e0f96d2a21e39d3ebaab85bbe1365780f7.

Reply:
Correction: my changes don't help in the described scenario. Checking the `prefix` allocation result before allocating `suffix` does help if there's a problem with the `prefix` allocation, but not if the `prefix` allocation went fine and the `suffix` allocation has a problem.

@andrewkozlik : are you aware of an elegant, generic approach to deal with this in cases of two or more `malloc()` calls? My goal with the  `RETURN_IF_NULL` macro was to make the code more readable and avoid error-prone boilerplate code, but I think there's no escaping of C's less-than-ideal lack of good error cleanup patterns. I wanted to avoid `goto cleanup;` patterns in the code, but since we can't teach the macro generically which pointers it has to free, I don't see good alternatives to it.

Reply:
The verbose solution is in https://github.com/trezor/trezor-firmware/pull/2449/commits/2c5fcf1ab4c011d9f186de5350f4e759f147ab98.

Reply:
Unfortunately I don't know of any pretty solution. `goto cleanup;` is fine, but I think we don't use it anywhere else. A sequence of `if (result == SUCCESS) {` blocks is sometimes used or nested `if` blocks would also work well, i.e.:
```c
int ret = 0;
char *out_buffer = malloc(outlen);
if (out_buffer != NULL) {
    char *in_buffer = malloc(inlen);
    if (in_buffer != NULL) {
        // code uses in_buffer and out_buffer here
        free(in_buffer);
    } else {
        ret = -1;
    }
    free(out_buffer);
} else {
    ret = -1;
}
return ret;
```
I'd leave it as is.